src/System.IO.FileSystem/src/System/IO/Win32FileSystemEnumerable.cs

```diff
@@ -1,83 +1,91 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-using System;
-using System.Collections;
 using System.Collections.Generic;
 using System.Diagnostics;
 using System.Diagnostics.Contracts;
-using System.Globalization;
 using System.Runtime.InteropServices;
-using System.Runtime.Versioning;
-using System.Security;
-using System.Text;
-using System.Threading;
 
-using Microsoft.Win32;
 using Microsoft.Win32.SafeHandles;
 
 namespace System.IO
 {
     // Overview:
     // The key methods instantiate Win32FileSystemEnumerableIterators. These compose the iterator with search result
-    // handlers that instantiate the FileInfo, DirectoryInfo, String, etc. The handlers then perform any
-    // additional required permission demands. 
+    // handlers that instantiate the FileInfo, DirectoryInfo, string, etc. The handlers then perform any
+    // additional required permission demands.
     internal static class Win32FileSystemEnumerableFactory
     {
-        internal static IEnumerable<String> CreateFileNameIterator(String path, String originalUserPath, String searchPattern,
+        internal static IEnumerable<string> CreateFileNameIterator(string path, string originalUserPath, string searchPattern,
                                                                     bool includeFiles, bool includeDirs, SearchOption searchOption)
         {
             Contract.Requires(path != null);
             Contract.Requires(originalUserPath != null);
             Contract.Requires(searchPattern != null);
 
-            SearchResultHandler<String> handler = new StringResultHandler(includeFiles, includeDirs);
-            return new Win32FileSystemEnumerableIterator<String>(path, originalUserPath, searchPattern, searchOption, handler);
+            SearchResultHandler<string> handler;
+
+            if (includeFiles && includeDirs)
+            {
+                handler = SearchResultHandler.FileSystemPath;
+            }
+            else if (includeFiles)
+            {
+                handler = SearchResultHandler.FilePath;
+            }
+            else if (includeDirs)
```

Comment:
Since this is an internal method and a problem with includeFiles / includeDirs would indicate a bug in our code rather than something we need to be throwing an exception for, instead of:
```C#
else if (includeDirs)
{
    handler = SearchResultHandler.DirectoryPath;
}
else
{
    Debug.Fail(...);
    throw ...;
}
```
how about just:
```C#
else
{
    Debug.Assert(includeDirs, "Should never be excluding both files and directories);
    handler = SearchResultHandler.DirectoryPath;
}
```
?